gcc/rust/backend/rust-compile.cc

```diff
@@ -198,13 +198,50 @@ CompileStructExprField::visit (HIR::StructExprFieldIdentifier &field)
 // Shared methods in compilation
 
 tree
-HIRCompileBase::coercion_site (tree rvalue, TyTy::BaseType *actual,
-			       TyTy::BaseType *expected, Location lvalue_locus,
+HIRCompileBase::coercion_site (tree rvalue, TyTy::BaseType *rval,
+			       TyTy::BaseType *lval, Location lvalue_locus,
 			       Location rvalue_locus)
 {
   if (rvalue == error_mark_node)
     return error_mark_node;
 
+  TyTy::BaseType *actual = rval;
+  if (actual->get_kind () == TyTy::TypeKind::PARAM)
+    {
+      TyTy::ParamType *p = static_cast<TyTy::ParamType *> (actual);
+      actual = p->resolve ();
+    }
+  else if (actual->get_kind () == TyTy::TypeKind::PLACEHOLDER)
+    {
+      TyTy::PlaceholderType *p = static_cast<TyTy::PlaceholderType *> (actual);
+      rust_assert (p->can_resolve ());
+      actual = p->resolve ();
+    }
+  else if (actual->get_kind () == TyTy::TypeKind::PROJECTION)
+    {
+      TyTy::ProjectionType *p = static_cast<TyTy::ProjectionType *> (actual);
+      actual = p->get ();
+    }
+
+  TyTy::BaseType *expected = lval;
+  if (expected->get_kind () == TyTy::TypeKind::PARAM)
+    {
+      TyTy::ParamType *p = static_cast<TyTy::ParamType *> (expected);
+      expected = p->resolve ();
+    }
+  else if (expected->get_kind () == TyTy::TypeKind::PLACEHOLDER)
+    {
+      TyTy::PlaceholderType *p
+	= static_cast<TyTy::PlaceholderType *> (expected);
+      rust_assert (p->can_resolve ());
+      expected = p->resolve ();
+    }
+  else if (expected->get_kind () == TyTy::TypeKind::PROJECTION)
+    {
+      TyTy::ProjectionType *p = static_cast<TyTy::ProjectionType *> (expected);
+      expected = p->get ();
+    }
```

Comment:
Maybe it would be worth to refactor this into a separate function if we want to reuse it later? Since if I understand correctly it's exactly the same code as above